src/python/grpcio/grpc/_cython/_cygrpc/aio/call.pxd.pxi

```diff
@@ -20,8 +20,9 @@ cdef class _AioCall:
         grpc_completion_queue * _cq
         grpc_experimental_completion_queue_functor _functor
         object _waiter_call
+        list _references
 
     @staticmethod
-    cdef void functor_run(grpc_experimental_completion_queue_functor* functor, int succeed)
+    cdef void functor_run(grpc_experimental_completion_queue_functor* functor, int succeed) with gil
     @staticmethod
-    cdef void watcher_call_functor_run(grpc_experimental_completion_queue_functor* functor, int succeed)
+    cdef void watcher_call_functor_run(grpc_experimental_completion_queue_functor* functor, int succeed) with gil
```

Comment:
Is this cosmetic change, or necessary change? I thought Cython functions by default requires GIL?

Reply:
TBH it was a surprise for me, with the previous code I suffered some random segfaults which they got solved automatically adding the `with gil`.

I look into the `uvloop` implementation code and all of the callbacks [1] were using explicitly the `with gil`, so I've added and the issues disappeared.

My rationale in front of this issue was, if we are saying explicitly that the C function of the library has to be called without keeping the GIL, we have to tell Cython that the callback - which could be executed immediately - needs to be executed acquiring the GIL yes or yes.

I guess that we would need to change the `CallbackWrapper`, I can take care of it if you want.


[1] https://github.com/MagicStack/uvloop/blob/7a4f00a3c22e52b88deb4b261e093e8affa4ea0a/uvloop/handles/tcp.pyx#L205

Reply:
Current version of this PR didn't add `with gil` to `CallbackWrapper`, but the tests still passing? This PR is using the Aio server, so theoretically, it suppose to still segfault?

I have seen that segfault before when I'm trying to get `EventManager` work, can you dig into which thread are the callbacks being ran? This issue might hit us again later.

Reply:
Ok, I will dig into it a bit more, in any case theoretically the initialization of the gRPC under Aio asks explicitly for not using any thread at all[1]  for running the callbacks.

[1] https://github.com/grpc/grpc/blob/master/src/python/grpcio/grpc/_cython/_cygrpc/aio/grpc_aio.pyx.pxi#L35

Reply:
Just double checked and seems that there are two threads, weird I would need to understand where this second thread comes from considering that the BT is only polluted of CPython functions and the way that Ive debug this was NOT USING the runner. So theoretically we should have only one thread (maybe unitest is running the tests in isolated tests?)

But in any case, without having the `with gil` in the callback, Cython does not know that you are requiring to acquire the GIL since you released when you called the gRPC syscall. Just doublechecked what Cython does when you specifically add the`with gil`, it adds the following call:

```
  #ifdef WITH_THREAD
  PyGILState_STATE __pyx_gilstate_save = __Pyx_PyGILState_Ensure();
  #endif
``` 

Seems that `PyGILState_Ensure` [1] is gonna make sure that you can safely call the C Python API (How I hate these tens of functions with similar semantics). TBH, I would say that 99% for sure that we should add this for the callbacks that are being called from the gRPC library.

If you need 100% confidence it will take a bit of more time

[1] https://docs.python.org/2/c-api/init.html#c.PyGILState_Ensure